lib/pacemaker/pcmk_sched_primitive.c

```diff
@@ -659,19 +658,58 @@ native_create_actions(pe_resource_t *rsc)
+    if ((current != NULL) && (chosen != NULL)
+        && (current->details != chosen->details)
+        && (rsc->next_role >= RSC_ROLE_STARTED)) {
+
+        pe_rsc_trace(rsc, "Moving %s from %s to %s",
+                     rsc->id, pe__node_name(current), pe__node_name(chosen));
+        is_moving = TRUE;
+
+        // Check whether resource is allowed to migrate
+        if (pcmk_all_flags_set(rsc->flags, pe_rsc_allow_migrate|pe_rsc_managed)
+            && !pcmk_any_flags_set(rsc->flags, pe_rsc_failed|pe_rsc_start_pending)
+            && !current->details->unclean && !chosen->details->unclean) {
+            allow_migrate = TRUE;
+        }
+
+        // This is needed even if migrating (though I'm not sure why ...)
+        need_stop = TRUE;
+    }
+
+    // Check whether resource is partially migrated and/or multiply active
+    if ((rsc->partial_migration_source != NULL)
+        && (rsc->partial_migration_target != NULL)
+        && allow_migrate && (num_all_active == 2)
         && (current->details == rsc->partial_migration_source->details)
         && (chosen->details == rsc->partial_migration_target->details)) {
-
-        /* The chosen node is still the migration target from a partial
-         * migration. Attempt to continue the migration instead of recovering
-         * by stopping the resource everywhere and starting it on a single node.
+        /* A partial migration is in progress, and the migration target remains
+         * the same as when the migration began.
          */
-        pe_rsc_trace(rsc, "Will attempt to continue with partial migration "
-                     "to target %s from %s",
-                     rsc->partial_migration_target->details->id,
-                     rsc->partial_migration_source->details->id);
+        pe_rsc_trace(rsc, "Partial migration of %s from %s to %s will continue",
+                     rsc->id, pe__node_name(rsc->partial_migration_source),
+                     pe__node_name(rsc->partial_migration_target));
+
+    } else if ((rsc->partial_migration_source != NULL)
+               || (rsc->partial_migration_target != NULL)) {
+        // A partial migration is in progress but can't be continued
+
+        if (num_all_active > 2) {
+            // The resource is migrating *and* multiply active!
+            crm_notice("Forcing recovery of %s because it is migrating "
+                       "from %s to %s and possibly active elsewhere",
+                       rsc->id, pe__node_name(rsc->partial_migration_source),
+                       pe__node_name(rsc->partial_migration_target));
+        } else {
+            // The migration source or target isn't available
+            crm_notice("Forcing recovery of %s because it can no longer "
+                       "migrate from %s to %s",
+                       rsc->id, pe__node_name(rsc->partial_migration_source),
+                       pe__node_name(rsc->partial_migration_target));
+        }
+        need_stop = TRUE;
```

Comment:
With this change, I think we're setting `need_stop = TRUE` in more cases than before. We're not setting `multiply_active = true` in some cases when we previously were, so we don't enter the `if (multiply_active)` block where we *selectively* set `need_stop` based on `rsc->recovery_type`. Instead, we set `need_stop = TRUE` and skip the `multply_active` block if `rsc->partial_migration_(source|target) != NULL`.

I wonder if this might have any unexpected effects that the regression tests simply don't catch when the recovery type is set to an unusual value like `block`. (Side note, we never did check for the `stop_only` recovery type...)

I haven't totally traced this through.

Reply:
The affected case is when a live migration was in progress but can't be continued -- which isn't really multiply active and shouldn't follow the multiple-active policy. Even if the resource is live migrating *and* multiply active, we need a full stop to abort the migration, so the multiple-active policy is irrelevant.

We don't need to check for stop_only because only a stop will be scheduled in that case. need_stop is needed when we have a stop and start scheduled for an active resource, to make the stop required, otherwise it looks like both (optional) actions are unnecessary since the current state is the end state.